### ISSUE TITLE
Support non-array data

### ISSUE THREAD

**Issue by ddeboer** [portphp/portphp#14]
Previously, the data-import relied heavily on arrays as data items. I think arrays are usually a good choice, as they are lightweight. However, I think this library should not make assumptions about the data structure that users are working with. 

The data is still limited to arrays in two places:
1. [Writers](https://github.com/portphp/portphp/blob/master/src/Writer.php#L22) typehint on arrays.
2. The [filters](https://github.com/portphp/portphp/tree/master/src/Filter) that ship with this library still expect arrays. 

(Previously we had ItemConverters that only supported arrays, but we replaced with the much more flexible [steps](https://github.com/portphp/steps) that don’t depend on arrays.)

What about just removing the array typehint everywhere? In writers that only support arrays, such as `ExcelWriter`, we will just throw an exception if the input is no array.

This continues the discussion from https://github.com/ddeboer/data-import/issues/208. I closed that issue, but I think it needs more careful consideration.

/cc @florianeckerstorfer

**Comment by sagikazarmark** [comment]
:+1:

**Comment by Baachi** [comment]
Definitively :+1: on this. Lets talk about the implementation. 

Here my thoughts about this:
I think it would be good, if we have an `Item` interface which have a get/set/has/remove method (or just extends the `ArrayInterface` and we rewrite all writers to use this item.

Then we can ship an `ArrayItem` class which acts as a proxy for arrays. And maybe we can also ship
a `GenericObjectItem` which use the reflection api to manipulate the data from an object.

**Comment by ddeboer** [comment]
Let’s not forget about Symfony’s PropertyAccess component, which we already use in the [ValueConverterStep](https://github.com/portphp/steps/blob/master/src/Step/ValueConverterStep.php#L40), for instance.

**Comment by sagikazarmark** [comment]
Sounds good. What about performance?

Also, should the reader return item instances? Or if it doesn't, we automatically convert it to one or throw exception?

**Comment by florianeckerstorfer** [comment]
Performance is probably a problem with PropertyAccess. The component has been developed for config files and templates, which can be compiled. For Plum I developed [Vale](https://github.com/cocur/vale), which is kinda a mix of PropertyAccess and [get_in](https://github.com/igorw/get-in) (+object traversal).

**Comment by ddeboer** [comment]
Vale looks interesting. Perhaps we could improve performance even more if we assume each data item has the same structure. We can then cache the data accessors used for the first data item that passed through a writer for subsequent ones.

**Comment by sagikazarmark** [comment]
I am pretty sure you can't assume this in every cases. Only if you look at converters, that can result in any number of data structures.

**Comment by ddeboer** [comment]
Good point. 

So then we have two options:

1. Assume arrays and throw an exception in case a non-array enters a writer. This makes users responsible for chaining the converters in the right way: when they work with data, their last step must convert the object into an array.

2. Use an intelligent but performant property accessor (which is optimised especially for arrays). 

2 seems the more reasonable thing to do. I’d like to compare Vale, get_in and PropertyAccess a bit before making a decision.

**Comment by sagikazarmark** [comment]
I think both makes sense. The array limitation probably makes it harder to use the library for edge cases when the item is not an array and someone want to use specific writers for that workflow.

Always using a property accessor also doesn't seem to be the right choice for me (at least not on workflow level). Since we are not limiting how data is sent through the workflow (by removing the array typehint from writers) it should not be limited how data is retrieved (again, at least not on workflow level).

I think introducing an item interface is a good idea. It allows us to work more easily with the data in filters and converters, however this is the case where accessing data is hard.

So I would say a constant property accessor usage would not be necessary for all items. What I would do is define a few access types for items: ArrayAccessable, ValeAccessable, etc. This way we can ensure if a filter/converter/writer can really access data of an item.

Also, readers should not be forced to return an item. Most current readers return arrays, which is just fine IMO. I would do a check in the workflow or add a reader wrapper for non-item-returning-readers. If the returned value is an array, an ArrayItem is created, if an object, and ObjectItem is created, if an Item, passed through the workflow. Accessor logic should be provided by the item.

I think this would give us great flexibility since the workflow itself doesn't know anything about the data, only the relevant parts of it requires specific features of it. However I am not quite sure about how it would perform with a massive amount of data. I am referring to the fact of having every item as an object, not just the use of property accessing logic.

**Comment by sagikazarmark** [comment]
What's the plan?

**Comment by ddeboer** [comment]
I like the Item objects. To still allow for plain arrays (in case of very large data sets) we could make each Item type implement ArrayAccess. That way, converters and writers can just continue to use the array notation for accessing data. But when the data consists of objects, the ValeAccessible item type will make the underlying data available in array notation. 

Unfortunately, PHP has no type hint for array || ArrayAccess, so if we don't want to create kludgy checks we would still have to remove the type hint.

**Comment by Baachi** [comment]
We don't need to use the property/accessor, we can just rely on the basic reflection api. This is the fastest option we can choose. I don't think we need a second library for this purpose.

**Comment by Baachi** [comment]
I also have a second approach we should consider. If we get an object, we can also convert it to an array, this library shows you how it works: https://github.com/avalanche123/Transform/.

But i would rather work with objects and not with arrays, because working with arrays need a huge amount of memory.

**Comment by ddeboer** [comment]
Are you saying arrays in PHP are less memory-efficient than objects? Do you have proof for that? :wink:

**Comment by Baachi** [comment]
Yes this is what i have to say. If you have an array and pass them to a function, php "copy" this array and pass the copy to this function. For objects there is a other behavior, php creates only a reference and pass them to the function.

Small script for benchmarking:
https://gist.github.com/Baachi/6d652f1f3174f5641eb9

**Comment by florianeckerstorfer** [comment]
@Baachi Only if the array does not contain references to other objects. Specially, what if an item contains a reference to the collection it is part of.

Using the Reflection API is slow, specially if you have to do it for every item.

**Comment by Baachi** [comment]
That's true, it becomes complicated if php have to deal with references in an array. 
We don't have to call the reflection every times, because we can `cache` some property access (or event better we can convert the object to an array one times).

There are different ways to go, we just need an unified api to get data from an item.

**Comment by sagikazarmark** [comment]
As I said earlier we probably shouldn't force Readers to return Item objects. A noticable number of them work with underlying iterators. What I would do is the following:

``` php
<?php

interface Item {}

interface ArrayAccessableItem extends Item
{
    /**
     * Returns the underlying data as an array
     *
     * Allows extraction logic even if the data itself is not stored as array
     *
     * @return array
     */
    public function toArray();
}

class ArrayItem implements ArrayAccessableItem, \ArrayAccess
{
    /**
     * @var array
     */
    private $data;

    /**
     * @param array $data
     */
    public function __construct(array $data)
    {
        $this->$data = $data;
    }

    /**
     * {@inheritdoc}
     */
    public function toArray()
    {
        return $this->data;
    }
}

class GenericObjectItem implements ArrayAccessableItem
{
    /**
     * @var object
     */
    private $data;

    /**
     * @param object $data
     */
    public function __construct($data)
    {
        $this->$data = $data;
    }

    /**
     * {@inheritdoc}
     */
    public function toArray()
    {
        return (array) $this->data;
    }
}

// From StepAggregator
// ...
foreach ($reader as $index => $item) {
    if (!$item instanceof Item) {
        if (is_array($item)) {
            $item = new ArrayItem($item);
        }
    }
}
// ...
```

One possible problem with this is that the Item object must know how to retrieve data from it, but I think it is necessary if we want to avoid bigger complexity.

**Comment by sagikazarmark** [comment]
And the writer part:

``` php
class DummyWriter implements Writer
{
	public function writeItem(Item $item)
	{
		if (!$item instanceof ArrayAccessableItem) {
			throw new \InvalidArgumenException('This writer needs the data in array form');
		}
	}
}
```

**Comment by ddeboer** [comment]
> if (!$item instanceof ArrayAccessableItem) {

But this checks on every item, which slows things down. Perhaps we should add some kind of `supportsItem()` method that is only called once? Of course, we then assume that all items are alike.

**Comment by sagikazarmark** [comment]
I still wouldn't force such thing on readers. Maybe we could have a separate `ItemReader` interface which indicates that the reader returns an item instance. Or we could always check the first behaviour and attach a middleware to the beggining of the stack so that it always gets converted.

I usually try to avoid `supports` and other methods like that because they add extra coupling.

**Comment by sagikazarmark** [comment]
ping @ddeboer @Baachi

**Comment by slashrsm** [comment]
Closing as not planned for the current maintenance line.

This RFC (non-array items / Item interface / property accessors) never reached an agreed implementation. After 10+ years and the 2.0 modernization, the public writer contract still uses `writeItem(array $item)`, and changing that would be a deliberate new major with a clear design + BC plan.

Users who need objects can convert to arrays in a step before writers that require arrays.

Feel free to reopen or open a fresh issue/PR if there is a concrete proposal targeting a future major.
